module/platform/OpenCR/HardwareIO/src/HardwareIO.cpp

```diff
@@ -0,0 +1,276 @@
+                log<NUClear::WARN>(fmt::format("OpenCR model information not recieved, restarting system"));
+                // Clear all packet queues just in case
+                queue_clear_all();
+                // Restart the system and exit the watchdog
+                startup();
+                return;
+            }
+
+            // Check what the hangup was
+            bool packet_dropped = false;
+            // The result of the assignment is 0 (NUgus::ID::NO_ID) if we aren't waiting on
+            // any packets, otherwise is the nonzero ID of the timed out device
+            while (uint8_t dropout_id = queue_item_waiting()) {
+                // delete the packet we're waiting on
+                packet_queue[dropout_id].erase(packet_queue[dropout_id].begin());
+                log<NUClear::WARN>(fmt::format("Dropped packet from ID {}", dropout_id));
+                // set flag
+                packet_dropped = true;
+            }
+
+
+            // Send a request for all servo packets, only if there were packets dropped
+            // In case the system stops for some other reason, we don't want the watchdog
+            // to make it automatically restart
+            if (packet_dropped) {
+                log<NUClear::WARN>("Requesting servo packets to restart system");
+                send_servo_request();
+            }
+        });
+
+        // When we receive data back from the OpenCR it will arrive here
+        // Run a state machine to handle reception of packet header and data
+        // If a packet is successfully emitted then we emit a StatusReturn message
+        on<IO>(opencr.native_handle(), IO::READ).then([this] {
+            // Process the response packet and emit a StatusReturn if applicable
+            handle_response();
+        });
+
+        on<Trigger<StatusReturn>, Sync<HardwareIO>>().then([this](const StatusReturn& packet) {
+            /* Error handling */
+
+            // Check we can process this packet
+            if (packet_queue.find(packet.id) == packet_queue.end()) {
+                log<NUClear::WARN>(fmt::format("Recieved packet for unexpected ID {}.", packet.id));
+                return;
+            }
+
+            // Check we're expecting the packet
+            if (packet_queue[packet.id].size() == 0) {
+                log<NUClear::WARN>(fmt::format("Unexpected packet data received for ID {}.", packet.id));
+                return;
+            }
+
+            /* All good now */
+
+            // Service the watchdog because we recieved a valid packet
+            emit<Scope::WATCHDOG>(ServiceWatchdog<HardwareIO>());
+
+            // Pop the front of the packet queue
+            auto& info = packet_queue[packet.id].front();
+            packet_queue[packet.id].erase(packet_queue[packet.id].begin());
+
+            // Check for packet errors
+            /// @todo Do we want to handle packets differently if they have errors?
```

Comment:
There is an issue open for OpenCR. Can still get rid of the todo comments in favour of the issue but I wouldn't say we should just do them, because this PR is already huge and I want to get this into main so Dex and I can work on issues concurrently in branches.